stl/src/atomic.cpp

```diff
@@ -5,32 +5,20 @@
 
 #include <yvals.h>
 
-#include <intrin.h>
-#pragma warning(disable : 4793)
+#include <atomic>
 
 _EXTERN_C
 
-// SPIN LOCK FOR shared_ptr ATOMIC OPERATIONS
-volatile long _Shared_ptr_flag;
+std::atomic_flag _Shared_ptr_flag{};
 
 _CRTIMP2_PURE void __cdecl _Lock_shared_ptr_spin_lock() { // spin until _Shared_ptr_flag successfully set
-#ifdef _M_ARM
-    while (_InterlockedExchange_acq(&_Shared_ptr_flag, 1)) {
-        __yield();
+    while (_Shared_ptr_flag.test_and_set(std::memory_order_acquire)) {

```

Comment:
Is it worth using a load spin-loop here while waiting for other threads to finish with the lock to avoid bouncing the cache line between threads?

Also, could the memory ordering issues be solved by putting a stronger memory order on the initial `test_and_set`?

```
// use acq/rel memory order on initial attempt to avoid reordering with prior unlock.
if (_Shared_ptr_flag.test_and_set(std::memory_order_acq_rel)) {
  do {
    // Read-only spin until we see other thread has released lock.
    do {
      _YIELD_PROCESSOR();
    } while (_Shared_ptr_flag.test(std::memory_order_relaxed));
  } while (_Shared_ptr_flag.test_and_set(std::memory_order_acquire)); // Use weaker order on subsequent attempts.
}
```

The added complexity of the code might outweigh any potential gains, though.
It probably depends how contended this lock will be, and may affect fairness of the lock.